include/etcpal/cpp/error.h

```diff
@@ -70,125 +75,852 @@ namespace etcpal
 ///   // Prints "Error: 'Out of memory'"
 /// }
 /// \endcode
+///
+/// Result is often used as the return type for functions; it has an implicit constructor from
+/// #etcpal_error_t to simplify return statements:
+/// \code
+/// etcpal::Result DoSomething()
+/// {
+///   // ... After we did something successfully...
+///   return kEtcPalErrOk;
+/// }
+/// \endcode
 class Result
 {
 public:
   Result() = delete;
   // Note: this constructor is not explicit by design, to allow implicit conversions e.g.
-  //   etcpal::Result res = etcpal_c_function_that_returns_error_t();
-  Result(etcpal_error_t code);
-  Result& operator=(etcpal_error_t code);
+  //   etcpal::Result FunctionReturningResult()
+  //   {
+  //     return kEtcPalErrOk;
+  //   }
+  constexpr Result(etcpal_error_t code) noexcept;
+  Result& operator=(etcpal_error_t code) noexcept;
 
-  bool IsOk() const;
-  etcpal_error_t code() const;
+  constexpr bool IsOk() const noexcept;
+  constexpr etcpal_error_t code() const noexcept;
   std::string ToString() const;
-  const char* ToCString() const;
+  const char* ToCString() const noexcept;
 
-  explicit operator bool() const;
+  constexpr explicit operator bool() const noexcept;
 
-  static Result Ok();
+  static constexpr Result Ok() noexcept;
 
 private:
   etcpal_error_t code_{kEtcPalErrOk};
-  const char* str_{nullptr};
-
-  void SetCode(etcpal_error_t code);
 };
 
 /// \brief Construct a Result from an error code.
-inline Result::Result(etcpal_error_t code)
+constexpr Result::Result(etcpal_error_t code) noexcept : code_(code)
 {
-  SetCode(code);
 }
 
 /// \brief Assign an error code to this Result.
-inline Result& Result::operator=(etcpal_error_t code)
+inline Result& Result::operator=(etcpal_error_t code) noexcept
```

Comment:
How can IsOk(), etc, be constexpr if this isn't constexpr?

Reply:
They are unrelated. I think the assignment operator theoretically _could_ be constexpr (I'm not sure it would satisfy the C++11 requirements) but I don't think it would do much good. As I understand it, it helps most in the constructor and observers, so that the compiler could evaluate an expression like `Result(kEtcPalErrOk).IsOk()` to simply `true` at compile-time.

Reply:
Ah, I see.  I was thinking about constexpr as something functional that worked like const and was inviolate.  Instead, it's more like inline -- it's optional and the compiler can use it as a constexpr if it determines that the entire expression/operation can be evaluated at compile time.